### PR TITLE
improve: close a campaign message when user has changed, fix toAfterDelay clause (SDKCF-3135)

### PR DESCRIPTION
**Pull request by imaginaris**
# Description
Displayed campaign will be closed if user logs out or changes to another user.
Campaign message view injection logic has been moved from BaseView to Router.
There is also a fix for `toAfterDelay` Nimble expression which didn't evaluate expectation properly.
DependencyManager's resolver property access is now synchronised because of intermittent runtime exceptions during unit test execution.

## Links
SDKCF-3135

# Checklist
- [X] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [X] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data **before every commit**, including API endpoints and keys
- [X] I ran `fastlane ci` without errors

### CODE REVIEW

File: RInAppMessaging/Classes/Router.swift [hunk 1]
```diff
@@ -24,6 +27,17 @@ internal class Router: RouterType {
         self.dependencyManager = dependencyManager
     }
 
+    func discardDisplayedCampaign() {
+        DispatchQueue.main.async {
```
Comment (malayli) [inline RInAppMessaging/Classes/Router.swift:31]: Is it because discardDisplayedCampaign is not executed on the main thread? or is it to fix UI issues?

**Review by malayli — APPROVED** [review]

File: RInAppMessaging/Classes/Router.swift (hunk 1 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:31]: it's not executed on main thread. I think I event got a runtime warning in the xcode

File: RInAppMessaging/Classes/Router.swift [hunk 2]
```diff
@@ -75,10 +89,57 @@ internal class Router: RouterType {
...
                 }
-                view.show(accessibilityCompatible: self.accessibilityCompatibleDisplay, onDismiss: {
+                guard let rootView = self.getKeyWindow(),
+                      self.findPresentedIAMView(from: rootView) == nil else {
+                    return
+                }
+
+                var parentView: UIView = rootView
+
+                // For accessibilityCompatible option, campaign view must be inserted to
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Router.swift:99]: The `accessibilityCompatible` stuff might be clearer if it was isolated in a separate function.

File: RInAppMessaging/Classes/Router.swift [hunk 3]
```diff
@@ -75,10 +89,57 @@ internal class Router: RouterType {
...
+                // UIWindow's main subview. Private instance of UITransitionView
+                // shouldn't be used for that - that's why it's omitted.
+                if self.accessibilityCompatibleDisplay,
+                    let mainSubview = rootView.subviews.first(
+                        where: { !$0.isKind(of: NSClassFromString("UITransitionView")!) }) {
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Router.swift:104]: This force unwrap seems potentially unsafe. Can we avoid it?

File: RInAppMessaging/Classes/Router.swift [hunk 4]
```diff
@@ -75,10 +89,57 @@ internal class Router: RouterType {
...
+
+                view.show(accessibilityCompatible: self.accessibilityCompatibleDisplay,
+                          parentView: parentView,
+                          onDismiss: {
                     completion(false)
                 })
             }
         }
     }
+
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Router.swift:116]: Could the below helper funcs be refactored into unit-testable UIView/UIWindow extensions?

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/// This wrapper ensures synchronized access to the value only for getter and setter.
+/// Mutating functions, subscript, incrementation etc. are not synchronized.
+@propertyWrapper
+struct AtomicGetSet<Value> {
+    private let queue = PropertyQueueGenerator.spawnQueue(domain: "IAM.AtomicProperty")
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:5]: Does each struct needs its own queue?

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/// This wrapper ensures synchronized access to the value only for getter and setter.
+/// Mutating functions, subscript, incrementation etc. are not synchronized.
+@propertyWrapper
+struct AtomicGetSet<Value> {
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:4]: Cool! 👍 

Is this property wrapper unit testable? If so, it'd be great to have tests.

File: Tests/InAppMessagingModuleTests.swift [hunk 7]
```diff
@@ -246,6 +249,33 @@ class InAppMessagingModuleTests: QuickSpec {
...
                             }
+
+                            it("will discard displayed campaign if user logs out") {
+                                let preference = IAMPreferenceBuilder().setUserId("user1").build()
+                                iamModule.registerPreference(preference)
+                                iamModule.registerPreference(nil)
+                                expect(router.wasDiscardCampaignCalled).to(beTrue())
+                            }
+
+                            it("will discard displayed campaign if user's changed") {
```
Comment (donnie-jp) [inline Tests/InAppMessagingModuleTests.swift:260]: user's changed -> user is changed

**Review by donnie-jp — COMMENTED** [review]
Basically LGTM - just left a few comments to check.

File: Tests/InAppMessagingModuleTests.swift (hunk 7 above)
Comment (imaginaris) [reply Tests/InAppMessagingModuleTests.swift:260]: Ah wrong tense. Not surprised.

File: RInAppMessaging/Classes/Router.swift (hunk 2 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:99]: Done

File: RInAppMessaging/Classes/Router.swift (hunk 3 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:104]: Hmm, yes I think we can

File: RInAppMessaging/Classes/Router.swift (hunk 4 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:116]: Done.

Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:116]: Done.

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 6 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:4]: I tried to write some tests but I couldn't produce anything sensible

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:5]: To ensure independent access between different properties

File: RInAppMessaging/Classes/Router.swift (hunk 3 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Router.swift:104]: Hmm, yes I think we can

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 6 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:4]: I tried to write some tests but I couldn't produce anything sensible

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift:5]: To ensure independent access between different properties

**Comment by imaginaris** [comment]
> Basically LGTM - just left a few comments to check.

Fixes for those comments in https://github.com/rakutentech/ios-inappmessaging/pull/39

**Pull request merged**